### PR TITLE
VideoSwin Backbone

### PR DESCRIPTION
**Pull request by ushareng**

### CODE REVIEW

**Comment by ushareng** [comment]
The input variable names are not fixed yet. For image classifier backbones `image_shape` fits right, what should we keep it here `video_shape` or you want to go with generic one

**Comment by ushareng** [comment]
Also, in the classifier I have used imageclassifier, which needs to be updated to video classifier 
Shall I put that change in this PR or how?

**Comment by mattdangerw** [comment]
We can ignore nightly failure, but please for `./shell/format.sh` and `./shell/api_gen.sh` to fix the linter errors.

File: keras_nlp/src/models/video_swin/video_swin_backbone.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        - [Video Swin Transformer](https://arxiv.org/abs/2106.13230)
+        - [Official Code](https://github.com/SwinTransformer/Video-Swin-Transformer)
+    """
+
+    def __init__(
+        self,
+        include_rescaling=False,
+        image_shape=(32, 224, 224, 3),
+        embed_dim=96,
+        patch_size=[2, 4, 4],
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_backbone.py:90]: We should probably call the arguments that vary per layer either `layerwise_patch_size` or `stackwise_patch_size`, etc, for consistency with other models.

File: keras_nlp/src/models/video_swin/video_swin_backbone.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    References:
+        - [Video Swin Transformer](https://arxiv.org/abs/2106.13230)
+        - [Official Code](https://github.com/SwinTransformer/Video-Swin-Transformer)
+    """
+
+    def __init__(
+        self,
+        include_rescaling=False,
+        image_shape=(32, 224, 224, 3),
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_backbone.py:88]: If this model is capable of taking in different numbers of frames or different image sizes, this should probably be `(None, None, None, 3)`.

File: keras_nlp/src/models/video_swin/video_swin_backbone.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    ```
+
+    References:
+        - [Video Swin Transformer](https://arxiv.org/abs/2106.13230)
+        - [Official Code](https://github.com/SwinTransformer/Video-Swin-Transformer)
+    """
+
+    def __init__(
+        self,
+        include_rescaling=False,
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_backbone.py:87]: Remove defaults that are specific to a given model size. `patch_size`, `window_size`, `depths` and `num_heads` should all have no default here. The defaults will be in the individual preset configs.

File: keras_nlp/src/models/video_swin/video_swin_backbone.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        drop_path_rate=0.2,
+        depths=[2, 2, 6, 2],
+        num_heads=[3, 6, 12, 24],
+        qkv_bias=True,
+        qk_scale=None,
+        **kwargs,
+    ):
+
+        # === Functional Model ===
+
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_backbone.py:105]: remove empty newline

File: keras_nlp/src/models/video_swin/video_swin_backbone.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+@keras_nlp_export("keras_nlp.models.VideoSwinBackbone")
+class VideoSwinBackbone(Backbone):
+    """A Video Swin Transformer backbone model.
+
+    Args:
+        image_shape (tuple[int], optional): The size of the input video in
+            `(depth, height, width, channel)` format.
+            Defaults to `(32, 224, 224, 3)`.
+        include_rescaling (bool, optional): Whether to rescale the inputs. If
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_backbone.py:41]: format lines to 80 characters. This looks like we are going over.

File: keras_nlp/src/models/video_swin/video_swin_backbone.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            set to `True`, inputs will be passed through a `Rescaling(1/255.0)` layer
+            and normalize with mean=[0.485, 0.456, 0.406] and std=[0.229, 0.224, 0.225].
+            Defaults to `False`.
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_backbone.py:44]: I would remove the default annotations here. They easily go stale and they will show up in the keras.io docs without this.

File: keras_nlp/src/models/video_swin/video_swin_backbone_test.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import pytest
+from keras import ops
+
+from keras_nlp.src.models.video_swin.video_swin_backbone import (
+    VideoSwinBackbone,
+)
+from keras_nlp.src.tests.test_case import TestCase
+
+
+class VideoSwinSBackboneTest(TestCase):
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_backbone_test.py:28]: Rewrite this test case to follow the standard for other backbones. Use the common routines for testing backbones and saving etc. `run_backbone_test`

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 8]
```diff
@@ -0,0 +1,1015 @@
...
+# limitations under the License.
+import keras
+import numpy as np
+from keras import layers
+from keras import ops
+from keras import random
+
+
+def window_partition(x, window_size):
+    """Partitions a video tensor into non-overlapping windows of a specified size.
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:22]: 80 characters or less

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 9]
```diff
@@ -0,0 +1,1015 @@
...
+            height // window_size[1],
+            window_size[1],
+            width // window_size[2],
+            window_size[2],
+            channel,
+        ],
+    )
+
+    x = ops.transpose(x, [0, 1, 3, 5, 2, 4, 6, 7])
+    windows = ops.reshape(
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:64]: return right away

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 10]
```diff
@@ -0,0 +1,1015 @@
...
+            -1,
+        ],
+    )
+    x = ops.transpose(x, [0, 1, 4, 2, 5, 3, 6, 7])
+    x = ops.reshape(x, [batch_size, depth, height, width, -1])
+    return x
+
+
+def get_window_size(x_size, window_size, shift_size=None):
+    """Computes the appropriate window size and potentially shift size for Swin Transformer.
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:112]: fix line length for every docstring here (I'll stop commenting this one)

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 11]
```diff
@@ -0,0 +1,1015 @@
...
+                the formula:
+                    (depth - window_size[0]) // shift_size[0] + 1) *
+                    (height - window_size[1]) // shift_size[1] + 1) *
+                    (width - window_size[2]) // shift_size[2] + 1)
+        Each element (attn_mask[i, j]) represents the attention weight between
+        window i and window j. A value of -100.0 indicates high negative attention
+        (preventing information flow), 0.0 indicates no mask effect.
+    """
+
+    img_mask = np.zeros((1, depth, height, width, 1))
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:185]: what's the overall structure of this mask? we likely want to find a way to rewrite this whole block using `ops` and not `numpy` and for loops.

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 12]
```diff
@@ -0,0 +1,1015 @@
...
+
+    def __init__(
+        self, patch_size=(2, 4, 4), embed_dim=96, norm_layer=None, **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.patch_size = patch_size
+        self.embed_dim = embed_dim
+        self.norm_layer = norm_layer
+
+    def __compute_padding(self, dim, patch_size):
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:295]: Why no one underscore? Double underscore is usually reserved for python builtins

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 13]
```diff
@@ -0,0 +1,1015 @@
...
+        window_size,
+        num_heads,
+        qkv_bias=True,
+        qk_scale=None,
+        attn_drop_rate=0.0,
+        proj_drop_rate=0.0,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        # variables
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:448]: config? variables usually refers to trainable weights (these are not)

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 14]
```diff
@@ -0,0 +1,1015 @@
...
+        qk_scale=None,
+        drop_rate=0.0,
+        attn_drop_rate=0.0,
+        drop_path_rate=0.0,
+        activation="gelu",
+        norm_layer=layers.LayerNormalization,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        # variables
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:785]: same here, these aren'te variables

File: keras_nlp/src/models/video_swin/video_swin_layers.py [hunk 15]
```diff
@@ -0,0 +1,1015 @@
...
+        - [Video Swin Transformer GitHub](https://github.com/SwinTransformer/Video-Swin-Transformer)
+    """
+
+    def __init__(
+        self, hidden_dim, output_dim, drop_rate=0.0, activation="gelu", **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.output_dim = output_dim
+        self.hidden_dim = hidden_dim
+        self._activation_identifier = activation
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers.py:236]: I see this in a few places, we should not do it.

in init
`self.activation = keras.activations.get(activation)`

in get_config
`"activation": keras.activations.serialize(self.activation),`

File: keras_nlp/src/models/video_swin/video_swin_layers_test.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        output_shape = patch_embedding_model(input_array).shape
+        expected_output_shape = (1, 8, 8, 8, 96)
+        self.assertEqual(output_shape, expected_output_shape)
+
+    def test_patch_embedding_get_config(self):
+        patch_embedding_model = VideoSwinPatchingAndEmbedding(
+            patch_size=(4, 4, 4), embed_dim=96
+        )
+        config = patch_embedding_model.get_config()
+        assert isinstance(config, dict)
```
Comment (mattdangerw) [inline keras_nlp/src/models/video_swin/video_swin_layers_test.py:44]: rewrite all of these tests to match the repo style. We do not use `assert` anywhere.

**Comment by mattdangerw** [comment]
> Also, in the classifier I have used imageclassifier, which needs to be updated to video classifier
Shall I put that change in this PR or how?

Any way is fine, but it might be easiest to start with the backbone, and just add the task as a follow up.

Can you include some colab showing that we have correct numerics? E.g. load some reference implementation, assign the weights over, show our implementation matches?

**Review by mattdangerw — COMMENTED** [review]
Left some initial comments. Make sure to take time to check for the style of things in the repo and do some rewriting of the source code as we bring it over. We want to make sure this is a proper port that leaves us with a uniform UX for all models we have checked in.
